android/src/main/java/io/jari/fingerprint/FingerprintModule.java

```diff
@@ -14,171 +13,125 @@
-            exception.initCause(secEx);
-            promise.reject(exception);
-        } catch (Exception ex) {
-            promise.reject(ex);
+            IS_CANCELED = false;
+            CANCELLATION_SIGNAL = new CancellationSignal();
+            fpm.authenticate(
+                    null,
+                    0,
+                    CANCELLATION_SIGNAL,
+                    new FingerprintManagerCompat.AuthenticationCallback() {
+                        @Override
+                        public final void onAuthenticationError(
+                                final int errorCode,
+                                final CharSequence errorString
+                        ) {
+                            super.onAuthenticationError(errorCode, errorString);
+                            IS_CANCELED = CANCELLATION_SIGNAL.isCanceled();
+                            promise.reject(Integer.toString(errorCode), errorString.toString());
+                        }
+
+                        @Override
+                        public final void onAuthenticationHelp(
+                                final int helpCode,
+                                final CharSequence helpString
+                        ) {
+                            super.onAuthenticationHelp(helpCode, helpString);
+                            final WritableNativeMap writableNativeMap = new WritableNativeMap();
+                            writableNativeMap.putInt("code", helpCode);
+                            writableNativeMap.putString("message", helpString.toString());
+                            FingerprintModule.this.getReactApplicationContext().getJSModule(
+                                    DeviceEventManagerModule.RCTDeviceEventEmitter.class
+                            ).emit("onFingerprintAuthenticationHelp", writableNativeMap);
+                        }
+
+                        @Override
+                        public final void onAuthenticationSucceeded(
+                                final FingerprintManagerCompat.AuthenticationResult result
+                        ) {
+                            super.onAuthenticationSucceeded(result);
+                            promise.resolve(null);
+                        }
+
+                        @Override
+                        public final void onAuthenticationFailed() {
+                            super.onAuthenticationFailed();
+                            final WritableNativeMap writableNativeMap = new WritableNativeMap();
+                            writableNativeMap.putInt("code", FINGERPRINT_ACQUIRED_AUTH_FAILED);
+                            writableNativeMap.putString("message", "Invalid fingerprint");
+                            FingerprintModule.this.getReactApplicationContext().getJSModule(
+                                    DeviceEventManagerModule.RCTDeviceEventEmitter.class
+                            ).emit("onFingerprintAuthenticationFailed", writableNativeMap);
```

Comment:
Same as before, why emit this? Why not reject the promise? The promise only gets rejected when there's a SecurityException now....